### PR TITLE
Add/strStartsWith

### PR DESCRIPTION
**Pull request by rgadon107**
@hellofromtonya PR #13 creates a new branch `add/strStartsWith` and is ready for code review.

### CODE REVIEW

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	protected function setUp() {
+		parent::setUp();
+
+		require_once CENTRAL_HUB_ROOT_DIR . '/src/config-store/internals.php';
+	}
+
+	/*
+	 *  Test that a string starts with a character or substring.
+	 */
+	public function test_that_string_starts_with_a_character_or_substring() {
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php]: @rgadon107 Make sure you are naming test methods following this convention:

```
test_should_[expected_behavior]_when_[condition]
```

And the method's DocBlock should be clear too.  So for this test, it would be:

```
/**
 * Test str_starts_with() should return `true` when string starts with the given character or substring.
 */
public function test_should_return_true_when_string_starts_with() {
```

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+		$needle   = 'Hello';
+		$haystack = 'Hello_world!';
+		$actual   = str_starts_with( $haystack, $needle, $encoding = 'UTF-8' );
+
+		$this->assertTrue( $actual );
+		$this->assertSame( true, $actual );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php]: This whole test needs to be consolidated.

```php
$this->assertTrue( str_starts_with( 'Hello', 'Hello world!' );
```

And let's add more tests than just one.  Check for a single character or other substrings:

```php
$this->assertTrue( str_starts_with( 'Hello', 'Hello world!' );
$this->assertTrue( str_starts_with( 'Hello w', 'Hello world!' );
$this->assertTrue( str_starts_with( 'H', 'Hello world!' );
$this->assertTrue( str_starts_with( 'Cor', 'Cornerstone' );
```

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	}
+}
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php]: This test suite is missing tests for when the character or substring is not at the start of the string.  Remember to test both true and false conditions.  You want to capture all edge cases to ensure your code works as designed:

```php
/**
 * Test str_starts_with() should return `false` when string does not start with the given character or substring.
 */
public function test_should_return_false_when_does_not_start() {
    $this->assertFalse( str_starts_with( 'world!', 'Hello world!' );
    $this->assertFalse( str_starts_with( 'Hey', 'Hello world!' );
    $this->assertFalse( str_starts_with( 'Press', 'WordPress rocks!' );
   $this->assertFalse( str_starts_with( 'stone', 'Cornerstone' );
}
```

**Comment by rgadon107** [comment]
@hellofromtonya I revised the unit tests for PR#13 (Add/strStartsWith) in response to your earlier code review of 05/17/19.

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+		$this->assertFalse( str_starts_with( 'Hello World', 'hey!' ) );
+		$this->assertFalse( str_starts_with( 'Cornerstone', 'cor' ) );
+		$this->assertFalse( str_starts_with( 'Cornerstone', 'stone' ) );
+	}
+
+	/**
+	 * Test should return true when substring is empty string.
+	 */
+	public function test_should_return_true_when_substring_is_empty() {
+		$this->assertTrue( str_starts_with( 'Cornerstone', '' ) );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php]: @rgadon107 This one should not be true, as `Cornerstone` starts with a `C` and not an empty space.  When you run the unit tests, does this test pass or fail?  If it passes, then we need to look at the function to see why.

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+	/**
+	 * Test should return false or empty when substring is falsey.
+	 */
+	public function test_should_return_false_or_empty_when_substring_is_falsey() {
+		$this->assertFalse( str_starts_with( 'Hello World!', null ) );
+		$this->assertFalse( str_starts_with( 'Hello', 0 ) );
+		$this->assertFalse( str_starts_with( 'know the code', '0' ) );
+
+		$this->assertEmpty( str_starts_with( 'Hello World!', null ) );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php]: @rgadon107 The function only returns `true` or `false`.  Right?  It’s a Boolean value that gets returned.  Therefore, the tests should expect a Boolean back and not an empty.

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php (hunk 4 above)
Comment (rgadon107) [reply mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php]: @hellofromtonya I checked the documentation for the the PHP function [mb_substr](https://www.php.net/manual/en/function.mb-substr.php). The documentation says (in regards to the `$length` parameter): "If omitted or NULL is passed, extract all characters to the end of the string."

As you can see by the test, when I assign `$needle` an empty string in the function `str_starts_with()`, the function sets the variable `$needle_length` as the length of the entire string in `$haystack`. 

The API function `getAllKeysStartingWith( $starts_with )` will only work if it receives a valid string ( `$needle`) from `str_starts_with()`. I think that the function `str_starts_with()` should include a conditional to check for `$needle = ''`. If true, it should throw an Exception and message to inform the user that a string was expected and none was received. 

Does that make sense? Would you implement that approach?

Comment (hellofromtonya) [reply mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php]: >If true, it should throw an Exception and message to inform the user that a string was expected and none was received.

Just a note:  An exception informs the developer.  We don't want exceptions to inform site owners or visitors to the site.  If you use an exception, then we hope that a developer will handle and fix the problem.

How would you validate that the function has received a haystack and needle to compare?  The lengths of both would need to be more than 0.

Consider the following:

```php
function str_starts_with( $haystack, $needle, $encoding = 'UTF-8' ) {
	$needle_length = mb_strlen( $needle, $encoding );

	// Using string lengths, check if given a haystack and needle to compare.
	if ( $needle_length == 0 || mb_strlen( $haystack, $encoding ) == 0 ) {
		throw new InvalidArgumentException(
			sprintf( 'The haystack and needle cannot be empty. Given: haystack [%s] and needle of [%s].', $haystack, $needle )
		);
	}

	return ( mb_substr( $haystack, 0, $needle_length, $encoding ) === $needle );
}
```

Notice that the code is using the length of the strings given.  If either string has 0 characters, then the error is thrown.

Why not use a boolean check for falsey, e.g. 

```php
if ( ! $haystack || ! $needle ) {
```
??  Why not?  Because a string of `'0'` would flag as a falsey and throw the error. Consider that you want to check if a string starts with a 0.  You couldn't with the falsey check.

Hint: Try it for yourself to validate what happens when you pass a `"0"` as either the haystack or needle with the falsey check:

```php
vdump_and_die( ( '0' ) );
```

Given that falsey isn't going to work when needing to check if it starts with 0, we need a different way to check.  We use the string's length, i.e. number of characters.  If there are no characters, that means an empty string was passed.

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+	/**
+	 * Test should throw exception when string length empty.
+	 */
+	public function test_should_throw_exception_when_string_is_empty() {
+		$this->expectException( \InvalidArgumentException::class );
+		$this->expectExceptionMessage(
+			'The haystack and needle cannot be empty. Given: haystack [Cornerstone] and needle of [].'
+		);
+
+		str_starts_with( 'Cornerstone', '' );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php]: @rgadon107 Add an assertion to check that an exception is thrown when no haystack is given but a needle is and another assertion for when both are empty.  Add both of these in this test method.

Hint:  the excepted message is bound to the specific instance of invoking `str_starts_with`.  That means the one `expectExceptionMessage()`assertion on line 58 is bound to line 62.  

To add 2 more assertions, copy line 58-62 twice and then change the expected message and arguments.

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
+	/**
+	 * Test should return false or empty when substring is falsey.
+	 */
+//	public function test_should_return_false_when_substring_is_falsey() {
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php]: @rgadon107 Delete this test method instead of commenting it out.

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+		$this->assertFalse( str_starts_with( 'Hello World', 'hey!' ) );
+		$this->assertFalse( str_starts_with( 'Cornerstone', 'cor' ) );
+		$this->assertFalse( str_starts_with( 'Cornerstone', 'stone' ) );
+	}
+
+	/**
+	 * Test should throw exception when string length ($haystack or $needle) is empty.
+	 */
+	public function test_should_throw_exception_when_string_is_empty() {
+		$this->expectException( \Error::class );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php]: @rgadon107 The error class that should occur here is `InvalidArgumentException`.  Right?  [Look at the code added here](https://github.com/rgadon107/cornerstone/pull/13/files#diff-aff0ac70304193f0d42b37b100807912R112).  Make sure you run the tests before pushing the new commits, i.e. to ensure the tests run as you except.

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+		$this->expectErrorMessage(
+			'The haystack and needle cannot be empty. Given: haystack [Cornerstone] and needle of [\'\'].'
+		);
+
+		str_starts_with( 'Cornerstone', '' );
+
+		$this->expectException( \Error::class );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php]: @rgadon107 Same here, as it should be `InvalidArgumentException`.

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+		$this->expectErrorMessage(
+			'The haystack and needle cannot be empty. Given: haystack [\'\'] and needle of [WordPress].'
+		);
+
+		str_starts_with( '', 'WordPress' );
+
+		$this->expectException( \Error::class );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/config-store/strStartsWith.php]: @rgadon107 Same here, as it should be `InvalidArgumentException`.
